File: ComputacaoParalela/MultiThreading/teste.swift
/*
Para fazer com que o programa execute as threads em um único núcleo, 
você pode usar uma fila serial do Grand Central Dispatch (GCD) em vez de criar threads separadas. 
A fila serial garantirá que as tarefas sejam executadas sequencialmente em um único núcleo.
Neste código, substituímos a criação das threads por uma fila serial do GCD chamada "serialQueue". 
Em seguida, usamos o método async para adicionar as tarefas à fila, 
garantindo que elas serão executadas sequencialmente, uma após a outra, em um único núcleo. 
Por fim, utilizamos o método sync com a flag.barrier para aguardar a conclusão de todas as tarefas antes de imprimir a mensagem "Todas as tarefas foram concluídas". 
Dessa forma, garantimos que todas as tarefas sejam concluídas antes de encerrar o programa.
*/

import Foundation

func performTask(_ index: Int) {
    print("Iniciando tarefa \(index)")
    
    // Lógica da tarefa a ser executada
    for i in 1...5 {
        print("Tarefa \(index): Executando passo \(i)")
        // Simulação de processamento
        for _ in 1...1_000_000 { }
    }
    
    print("Concluindo tarefa \(index)")
}
let numberOfThreads = 1000 // Número de threads desejado (500)
var threads: [Thread] = []

for i in 0..<numberOfThreads {
    let thread = Thread {
        performTask(i)
    }
    threads.append(thread)
    thread.start()
}

// Verifica se todas as threads foram concluídas

var allThreadsCompleted = false
while !allThreadsCompleted {
    allThreadsCompleted = true
    for thread in threads {
        if thread.isExecuting {
            allThreadsCompleted = false
            break
        }
    }
    if !allThreadsCompleted {
        Thread.sleep(forTimeInterval: 0.1) // Aguarda 0.1 segundos
    }
}

print("Todas as tarefas foram concluídas")

File: ComputacaoParalela/MultiThreading/C/teste.c
/*
Neste código, utilizamos a biblioteca pthread.h para criar as threads. A função performTask representa a tarefa que será executada por cada thread. A função main cria as threads e aguarda sua conclusão utilizando a função pthread_join. Assim, garantimos que todas as tarefas serão concluídas antes de imprimir a mensagem "Todas as tarefas foram concluídas".
*/
#include <stdio.h>
#include <pthread.h>

#define NUMBER_OF_THREADS 1000

void *performTask(void *arg) {
    int index = *((int *)arg);
    printf("Iniciando tarefa %d\n", index);
    
    // Lógica da tarefa a ser executada
    for (int i = 1; i <= 5; i++) {
        printf("Tarefa %d: Executando passo %d\n", index, i);
        // Simulação de processamento
        for (int j = 0; j < 1000000; j++) { }
    }
    
    printf("Concluindo tarefa %d\n", index);
    return NULL;
}

int main() {
    pthread_t threads[NUMBER_OF_THREADS];
    int threadArgs[NUMBER_OF_THREADS];

    for (int i = 0; i < NUMBER_OF_THREADS; i++) {
        threadArgs[i] = i;
        int result = pthread_create(&threads[i], NULL, performTask, (void *)&threadArgs[i]);
        if (result != 0) {
            printf("Falha ao criar a thread %d\n", i);
            return 1;
        }
    }

    // Aguarda a conclusão de todas as threads
    for (int i = 0; i < NUMBER_OF_THREADS; i++) {
        pthread_join(threads[i], NULL);
    }

    printf("Todas as tarefas foram concluídas\n");
    return 0;
}

